### ISSUE TITLE
Transaction promise is mutating after being resolved (resolves to undefined after a while)

### ISSUE THREAD

**Issue by gajus** [tgriesser/knex#1052]
Detail description https://github.com/petkaantonov/bluebird/issues/863

**Comment by elhigu** [comment]
Verified bug, here is test case:

```js
test('#1052 - transaction promise mutating', function() {
  var transactionReturning = knex.transaction(function(trx) {
    return trx.insert({id: 1, name: 'A'}).into('test_table');
  });
  return Promise.all([transactionReturning, transactionReturning])
    .spread(function (ret1, ret2) {
      if (ret1 !== ret2) throw new Error("Promise was mutated when it was resolved twice.");
    });
});
```

**Comment by chuym** [comment]
Bluebird checks if the passed objects are instances of the `Promise` class in the `all` method on [this line](https://github.com/petkaantonov/bluebird/blob/master/src/promise_array.js#L21), otherwise it checks if it's a "Promise-like" class, (which `Transaction` is) and then does some failover functionality I haven't delved into yet. 

A quick workaround is to pass the `Promise` object and that should work:
````
test('#1052 - transaction promise mutating', function() {
  var transactionReturning = knex.transaction(function(trx) {
    return trx.insert({id: 1, name: 'A'}).into('test_table');
  });
  return Promise.all([transactionReturning._promise, transactionReturning._promise])
    .spread(function (ret1, ret2) {
      if (ret1 !== ret2) throw new Error("Promise was mutated when it was resolved twice.");
    });
});
````
I sent a fix on #1114 which makes `Transaction` a "Bluebird-like" promise object

**Comment by rkaw92** [comment]
This is really bad. Here is a test case which depicts just how dangerous this is:
```js
'use strict';

const knex = require('knex');
const connection = knex({
	client: 'sqlite3',
	connection: ':memory:'
});

const initPromise = connection.transaction(function(transaction) {
	// The body of the transaction probably does not matter:
	return transaction.schema.createTable('users', function(table) {
		table.text('userID');
		table.text('password');
	});
});

initPromise.then(function() {
	console.log('First .then() fulfilled');
	// Introduce the mutation:
	initPromise.then(function() {
		throw new Error('Promise overwriter');
	});

	setTimeout(function() {
		initPromise.then(function() {
			console.log('Second .then() fulfilled');
		}, function(error) {
			console.log('Second .then() ERROR: %s', error);
		});
	}, 2000);
});
```

Essentially, it makes any kind of code with lazy initialization (where the init promise is cached without wrapping / deriving a `then`) broken. As an aside, I wish we had an explicit execution chainable (`.do()`?), instead of pretending that queries/transactions are promises at all.

A workaround is to call `.then()` immediately and save the result of it as the "singleton promise", or otherwise wrap the transaction in `when()`, `Promise.resolve()` etc., which essentially has the same effect.

I believe it is essential for this idiosyncrasy to be documented (i.e. do not call `.then()` many times on the same Transaction object), because the problems that it causes are __extremely confusing__. In my case, a database abstraction class which had 2 methods (`get` and `add`) would reject with errors from `get`, but those errors could only ever be thrown from `add`, and it turned out that the "lazy init" promise (waited on by both methods) was somehow mutating its resolution status from `add`, "inheriting" its rejection.

**Comment by elhigu** [comment]
@rkaw92 I can't believe this has been left unresolved and forgotten. I'll make sure this get fixed for 0.13 release.

**Comment by rkaw92** [comment]
@elhigu From looking at the code, this seems especially suspicious: https://github.com/tgriesser/knex/blob/master/src/transaction.js#L261

It essentially replaces the promise reference with another. However, this means that by calling any Promise method (`.then()`) on the `Transaction`, one is always replacing the previous promise, and subsequent methods are executed on a *moving target* - different promise instances.

One thing I couldn't help but notice is the comment in line 258, which refers to the "current Target". The "Target" is a term used in `/src/interface.js`, which makes me think that this is some ancient part of `interface` which had been copied into `transaction` and has since escaped refactoring / debugging.
